
# Genome Size Estimation
Raw genomic reads were filtered for contaminant and mitochondria by aligning to contaminant genomes using bwa mem and discarding aligned reads using pysam. Reads were quality and adapter trimmed using fastp using default parameters for single end reads, and the '--detect_adapter_for_pe' flag for paired end reads. Remaining reads were kmer-counted using jellyfish and a kmer size of 31. These counts were made into a histogram with kat hist (kmer analysis toolkit). Genome sizes were estimated using genomescope, using the mean read length as reported by fastp, the kmer size of 31, and the kmer frequency spectra as produced by jellyfish. Genome sizes were only reported for genomes that had sufficient reads to meet 25X coverage of the genome, as suggested by genomescope user manual. Furthermore, The kmer frequency spectra for any genome with at least 25X coverage was manually inspected for a poisson distribution shape. If the spectra did not fit a poisson distribution, the estimated genome size from genomescope was not reported.  