---
title: "Copy Number Variation"
author: "Kirsten Gotting"
date: "`r Sys.Date()`"
output:
  html_document:
    theme: yeti
    highlight: kate
    code_folding: show
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE, error = FALSE)
knitr::opts_knit$set(root.dir = '..')
library("googlesheets4")
```



```{r def-env}

library(pheatmap)
library(RColorBrewer)
library(pvclust)
library(vegan)
library(UpSetR)
source('scripts/color_palettes.R')
library(cowplot)
library(ggridges)
library(tidyverse)


source('scripts/working_genomes.R')

length(genomes_keep)

make_ordered_dataframe <- function(data, order, col){
  datalist = list()
  for (t in order) {
    ani_vals <- c(data[grep(t, data$g1), col, drop = TRUE],
                  data[grep(t, data$g2), col, drop = TRUE])
    genome_vals <- c(data[grep(t, data$g1), ]$g2, 
                     data[grep(t, data$g2), ]$g1)
    if(length(ani_vals > 0)){
      data <- data %>% filter(g1 != t, g2 != t)
      datalist[[t]] <- data.frame(g1 = t, 
                                  g2 = genome_vals, 
                                  dn_ds = ani_vals)
    } else {
      datalist[[t]] <- data.frame()
    }
  }
  data <- bind_rows(datalist)
  data$g1 <- factor(data$g1, levels = order)
  data$g2 <- factor(data$g2, levels = order)
  invisible(data)  
}

esc <- c(           "ICBG2046",
                    "ICBG2048",
                    "ICBG2047",
                    "ICBG2049",
                    "ICBG712",
                    "ICBG721",
                    "ICBG1054",
                    "ICBG726",
                    "ICBG1065",
                    "ICBG1075",
                    "NIGD00000000",
                    "ICBG710",
                    "ICBG730",
                    "ICBG751",
                    "ICBG733",
                    "ICBG1096",
                    "ICBG742",
                    "NIGB00000000",
                    "LGSR00000000",
                    "NQYS00000000",
                    "ICBG731",
                    "ICBG736",
                    "NIGC00000000",
                    "NQYR00000000",
                    "NQYQ00000000")

munge_gca <- function(x){
  paste(strsplit(x, split = "_")[[1]][c(1,2)], collapse = "_")
}

ant_ag <- c("Lower",
            "Coral",
            "Higher", 
            "Leafcutter")
tool_sub <- c('virulence', 'led', 'MEROPS', 'cazy', 'CARD', 'busco')
tools_name <- c('virulence' = 'Virulence Genes', 
                'led' = 'Lipase Genes', 
                'MEROPS' = 'Peptidase Genes', 
                'cazy' = 'CAZyme Genes', 
                'CARD' = 'Resistance Genes', 
                'busco' = 'Single Copy Genes')

```

```{r load-data}

load('rdata/orthologues_long.rdata')

metadata <- read_csv('tables/metadata.csv') %>% 
  mutate(genome = sub("\\..*$", "", acc)) %>%
  mutate(Agriculture = ifelse(!clade_groups == 'Escovopsis', yes = clade_groups, no = Agriculture)) %>%
  mutate(Agriculture1 = ifelse(!clade_groups == 'Escovopsis', yes = clade_groups, no = Agriculture1)) %>%
  mutate(acc = sub("\\..*$", '', acc)) 

orthog_meta <- orthog_long %>%
  mutate(acc = ifelse(grepl("SPDT00000000", acc), yes = "GCA_008477525.1", no = acc)) %>%
  mutate(acc = sub("\\..*$", '', acc)) %>%
  group_by(Orthogroup) %>%
  mutate(n_genes = length(unique(genes))) %>%
  ungroup() %>%
  group_by(acc, Orthogroup) %>%
  mutate(n_genes_genome = length(unique(genes))) %>%
  filter(n_genes_genome <= 10) %>%
  ungroup() %>%
  select(-genome) %>%
#  mutate(g1 = sub("\\..*$", "", genome)) %>%
#  mutate(genome = sub("\\..*$", '', genome)) %>%
  left_join(., metadata, by = 'acc') %>%
  dplyr::filter(genome != 'GCA_000225605')

orthog_meta %>% group_by(clade_groups) %>%
  summarize(mean_copy_number = mean(n_genes_genome))

merops_fams <- read_tsv('tables/merops_family_map.txt', col_names = c('annotation', 'merops_family')) %>%
  distinct()

merops_fam_ls        <- merops_fams$merops_family
names(merops_fam_ls) <- merops_fams$annotation

annot    <- read_tsv('tables/all_annotations.txt', col_names = c('genome', 'gene', 'tool', 'annotation')) %>%
  filter(tool %in% tool_sub, annotation != '-') %>%
  mutate(genome = sub("\\..*$", "", genome)) %>%
  mutate(annotation = ifelse(tool == 'MEROPS',
                             yes = sub("-.*$", "", annotation),
                             no = annotation)) %>%
  mutate(annotation = ifelse(tool == 'MEROPS',
                             yes = ifelse(annotation %in% names(merops_fam_ls), yes = merops_fam_ls[annotation], no = annotation), ## integrate family ids, but keep undefined families
                             no = annotation)) %>%
  filter(!grepl('mix.component', annotation)) %>%
  mutate(annotation = ifelse(tool == 'virulence',
                             yes = sub("_.*$", "", annotation),
                             no = annotation))



```

```{r get_bgc_data}
files <- list.files("tables/bigscape_parsed_networks", 
                    pattern = 'network', 
                    recursive = TRUE, 
                    full.names = TRUE)




all_bgc <- lapply(files, function(x){
  table <- read_tsv(x, col_names = c('component', 'bigscape')) %>%
    mutate(BGC_type = sub("_.*$", "", basename(x)))
}) %>% bind_rows() %>%
  group_by(bigscape) %>%
  mutate(acc = case_when(
    grepl('GCA', bigscape) ~ munge_gca(bigscape),
    grepl('SPDT', bigscape) ~ "GCA_008477525.1",
    grepl('BGC', bigscape) ~ bigscape,
    TRUE ~ sub("_.*$", "", bigscape)
  )) %>%
  ungroup() %>%
  mutate(acc = sub("\\..*$", '', acc)) %>%
  left_join(., metadata, by = 'acc') %>%
  mutate(Presence = 1) %>%
  group_by(component, BGC_type) %>%
  mutate(num_agricultures = length(unique(Agriculture[which(Agriculture %in% ant_ag)]))) %>% ## removes outgroups from count
  filter(!is.na(acc)) %>%
  mutate(Agriculture = ifelse(is.na(Agriculture),
                              yes = 'Outgroup',
                              no = Agriculture)) %>%
  ungroup() %>%
  group_by(BGC_type, component) %>%
  mutate(bgc_plot_label = ifelse(any(grepl('BGC', acc)),
                                 yes = paste(grep('BGC', acc, value = TRUE), collapse = ","),
                                 no = paste0(BGC_type, "_", component))) %>%
  ungroup() %>%
  group_by(component) %>%
  mutate(n_genomes = length(unique(acc))) %>%
  ungroup() %>%
  group_by(acc) %>%
  mutate(n_BGC = length(unique(component))) %>%
  ungroup() %>%
  filter(!BGC_type %in% c('mix'),
         !grepl('BGC', acc)) %>% ## only look at BGC that are in at least 1 ant agriculture
#  select(bgc_plot_label, acc, Presence, BGC_type) %>%
  mutate(bgc_plot_label = sub('-', '.', bgc_plot_label)) %>% 
#  select(-BGC_type) %>% 
  select(component, acc, clade_groups, Presence, bgc_plot_label, Agriculture, BGC_type) %>%
  mutate(tool = 'BGC') %>%
  group_by(acc, component) %>%
  mutate(n_genes_genome = length(component)) %>% ## count the copies of each BGC
  ungroup() %>% 
  rename('Orthogroup' = component) %>%
  rename('annotation' = bgc_plot_label) %>%
  select(-Presence) %>%
  distinct()






```







```{r munge-data-count-copies}
count_increase = orthog_meta %>%
  select(Orthogroup, n_genes_genome, clade_groups) %>%
  group_by(clade_groups, Orthogroup) %>%
  summarize(med_genes = median(n_genes_genome)) %>%
  spread(clade_groups, med_genes) %>%
  group_by(Orthogroup) %>%
  mutate(ratio = Escovopsis/mean(c(Hypomyces_Cladobotryum, Trichoderma))) 

## more than 1 == copy number increase

count_increase %>% filter(ratio > 1) %>% nrow()

## less than 1 == copy number decrease

count_increase %>% filter(ratio < 1) %>% nrow()

orthog_meta %>% 
    select(Orthogroup, genome, clade_groups, n_genes_genome) %>%
    distinct() %>%
    ggplot(aes(x = clade_groups, y = n_genes_genome)) +
    geom_violin() +
    theme_classic() +
    labs(y = 'Gene Copy Number', 
         title = 'Copy Number Distribution')
  










orthog_long <- orthog_long %>%
  mutate(genome = sub("\\..*$", "", genome)) %>%
  filter(genome != "GCA_000225605") %>% ## remove cordycep genome for this analysis 
  rename('gene' = genes) %>%
  select(-acc) %>%
  mutate(genome = ifelse(genome == "SPDT00000000", yes = 'GCA_008477525', no = genome))



orthog_meta_ls <- lapply(tool_sub, function(x){
  if(x == 'led'){
    annot_sub <- annot %>% filter(tool == x) %>%
      group_by(genome, gene) %>%
      mutate(annotation = list(strsplit(annotation, split = ',')[[1]])) %>%
      ungroup() %>%
      unnest_longer(annotation)
  } else {
    annot_sub <- annot %>% filter(tool == x)}
  ## 90% or higher of the genes in the orthogroup need to be annotated to one of the tools to be included 
  orthog_meta <- orthog_long %>%
    inner_join(., annot_sub, by = c('genome', 'gene')) %>%
    mutate(metadata = ifelse(gene %in% annot_sub$gene, yes = TRUE, no = FALSE)) %>%
    group_by(Orthogroup) %>%
    mutate(n_genes = length(unique(gene))) %>%
    mutate(n_genes_in = length(which(metadata))) %>%
    ungroup() %>%
    group_by(genome, Orthogroup) %>%
    mutate(n_genes_genome = length(unique(gene))) %>%
    ungroup() %>%
    mutate(prop_genes_in = (n_genes_in/n_genes)*100) %>%
    mutate(g1 = sub("\\..*$", "", genome)) %>%
    filter(prop_genes_in >= 90) %>%
    mutate(genome = sub("\\..*$", '', genome)) %>%
    left_join(., metadata, by = 'genome') %>%
    mutate(tool = tools_name[x])
  if(x == 'cazy'){
    orthog_meta <- orthog_meta %>% filter(annotation != 'N')
  }
  orthog_meta
}) %>% bind_rows() %>%
      select(Orthogroup, acc, clade_groups, n_genes_genome, tool, annotation, Agriculture) %>%
  mutate(acc = sub("\\..*$", "", acc)) %>%
  bind_rows(., select(all_bgc, -BGC_type)) %>%
  mutate(tool = sub(" Genes", "", tool))



## mean copies of orthogroups
orthog_meta_ls %>% group_by(clade_groups, tool) %>%
    summarize(mean_copy_number = mean(n_genes_genome)) %>%
  spread(tool, mean_copy_number) %>% View()

orthog_meta_ls$tool <- factor(orthog_meta_ls$tool, levels = c("Single Copy",
                                                              "Resistance",
                                                              "Virulence",
                                                              "Lipase",
                                                              "Peptidase",
                                                              "CAZyme", "BGC"))

tools <- as.character(unique(orthog_meta_ls$tool))

```




```{r fig-3a}
###########################
## Making scatter plots ###
###########################


df <- orthog_meta_ls %>%
  select(Orthogroup, n_genes_genome, clade_groups, acc, tool) %>%
  distinct() %>%
  mutate(clade_groups = ifelse(clade_groups %in% c('Trichoderma', 'Hypomyces_Cladobotryum'), 
                               yes = 'Outgroup',
                               no = clade_groups)) %>%
  group_by(clade_groups, tool, Orthogroup) %>%
  summarize(mean = median(n_genes_genome)) %>%
  spread(clade_groups, mean) %>%
  replace_na(list(Escovopsis = 0, Outgroup = 0)) %>%
  group_by(tool, Escovopsis, Outgroup) %>%
  mutate(NumberGenes = length(Orthogroup)) %>%
  ungroup() %>% 
  select(-Orthogroup) %>%
  distinct() %>%
  mutate(tool = sub(" Genes", "", tool))


ratios_df <- lapply(tools, function(x){
## loss
a <- df %>% dplyr::filter(tool == x) %>%
  filter(Escovopsis == 0 , Outgroup > 0) %>%
  summarize(value = sum(NumberGenes)) %>%
  mutate(tool = x) %>%
  mutate(type = 'Loss')

## GAIN
b <- df %>% dplyr::filter(tool == x) %>%
  filter(Escovopsis > 0 , Outgroup == 0) %>%
  summarize(value = sum(NumberGenes)) %>%
  mutate(tool = x) %>%
  mutate(type = 'Gain')
bind_rows(a, b) %>% spread(type, value) %>%
  mutate(ratio = Loss/Gain) %>%
  arrange(ratio)


}) %>% bind_rows() %>%
  mutate(tool = sub(" Genes", "", tool))


df <- df %>% 
  left_join(., ratios_df, by = 'tool') %>%
  mutate(label = paste0(tool, " (", Loss, ":", Gain, ")"))


ratios_df

colors_heat <- colorRampPalette(c("#132B43", "#56B1F7"))



df$label <- factor(df$label, levels = unique(df$label)[c(4, 5, 6, 2, 3, 7, 1)])




fig2a <- df %>%
  filter(tool != 'Single Copy') %>%
  ggplot() +
  geom_point(aes(x = Outgroup, y = Escovopsis, size = NumberGenes)) +
  #geom_tile(aes(x = Outgroup, y = Escovopsis, fill = NumberGenes)) +
  geom_abline(intercept = 0, slope = 1, linetype = "dashed", colour = 'grey') +
  facet_wrap(~label, scales = 'free') +
  theme_classic() +
  labs(y = 'Median Copy Number in Escovopsis',
       x = 'Median Copy Number in Outgroups') +
  scale_x_continuous(breaks = seq(0, max(df$Outgroup), by = 2),
                     labels = seq(0, max(df$Outgroup), by = 2)) +
  scale_y_continuous(breaks = seq(0, max(df$Escovopsis)),
                     labels = seq(0, max(df$Escovopsis))) +
  theme(legend.position = 'bottom',
        text = element_text(size = 16),
        axis.text.x = element_text(angle = 45))

fig2a


```

```{r fig3b-all}
##################################################################
#### TRYING DIVERSITY METRICS 
#################################################################


x = tools[7]

div_mats <- lapply(tools, function(x){
  print(x)
  div_mat <- orthog_meta_ls %>% 
    select(acc, tool, annotation) %>%
    filter(tool == x) %>%
    select(-tool) %>%
    group_by(acc, annotation) %>%
    mutate(n_annotation = length(annotation)) %>%
    ungroup() %>%
    distinct() %>%
    spread(annotation, n_annotation) %>%
    column_to_rownames(var = 'acc')
  
  div_mat[is.na(div_mat)] <- 0
  div_mat = div_mat[, colSums(div_mat) >= 1, ] ## make sure there's at least one gene copy
  as.matrix(div_mat)
})

names(div_mats) <- tools

div_out <- lapply(names(div_mats), function(x){
  print(x)
  div_mat <- div_mats[[x]]
  H <- diversity(div_mat) ## shannon index
  evenness <- H/log(specnumber(div_mat)) ## evenness
  simp <- diversity(div_mat, "simpson")
  invsimp <- diversity(div_mat, "inv")
  S <- specnumber(div_mat) # observed number of species
  
  cbind(S, H, evenness, simp, invsimp) %>% 
    data.frame() %>%
    rownames_to_column(var = 'genome') %>%
    mutate(tool = x)
}) %>%
  bind_rows() %>%
  left_join(., metadata, by = 'genome')

div_out$Agriculture <- factor(div_out$Agriculture, levels = rev(c("Trichoderma",
                                                                                "Hypomyces_Cladobotryum",
                                                                                "Lower",
                                                                                "Coral",
                                                                                "Higher",
                                                                                "Leafcutter"
)))







plot_ls <- lapply(unique(div_out$tool), function(x){
  div_out %>% select(Agriculture, H, tool, S, evenness) %>%
    gather(metric, value, -tool, -Agriculture) %>%
    mutate(metric_n = case_when(
      metric == 'H' ~ 'Shannon Diversity',
      metric == 'S' ~ 'Number of Genes',
      metric == 'evenness' ~ 'Evenness'
    )) %>% 
    filter(tool == x) %>%
    ggplot(., aes(x = Agriculture, y = value, color = Agriculture, group = Agriculture)) +
    geom_boxplot() +
    scale_color_manual(values = colors[na.omit(unique(metadata$Agriculture))]) +
    theme_classic() +
    theme(legend.position = 'none') +
    labs(y = '', x = 'Grouping', title = x) +
    facet_wrap(~ metric_n, scales = 'free') +
    theme(axis.text.x = element_blank(), axis.ticks.x = element_blank())
})



all <- cowplot::plot_grid(plotlist = plot_ls, labels = 'AUTO', ncol = 1)
outfile = 'plots/gene_copy_box_diversity_all_metrics.pdf'
save_plot(plot = all, filename = outfile, base_height = 20, base_width = 8.6)
outfile = 'plots/gene_copy_box_diversity_all_metrics.png'
save_plot(plot = all, filename = outfile, base_height = 20, base_width = 8.6)

div_out$tool <- factor(div_out$tool, levels = unique(div_out$tool))



div_out$tool <- factor(div_out$tool, levels = c('Lipase', 'Peptidase', 'CAZyme',  'Resistance', 'Virulence', "BGC", 'Single Copy'))

fig2b <- div_out %>% 
  filter(tool != 'Single Copy') %>%
  ggplot(., aes(x = S, y = H, color = Agriculture)) +
  geom_jitter(size = 2) +
  scale_color_manual(values = colors[na.omit(unique(metadata$Agriculture))]) +
  theme_classic() +
  labs(y = 'Entropy', x = 'Number of Genes') +
  facet_wrap(~tool, scales = 'free') +
  theme(legend.position = 'bottom',
        text = element_text(size = 16),
        axis.text.x = element_text(angle = 45)) 



#fig2a

#fig2b

cowplot::plot_grid(plotlist = list(fig2a, fig2b), labels = 'AUTO') %>%
  ggsave2( filename = 'plots/figure4.png', height = 8, width = 14)

cowplot::plot_grid(plotlist = list(fig2a, fig2b), labels = 'AUTO') %>%
  ggsave2( filename = 'plots/figure4.pdf', height = 8, width = 14)


## SCG plots


scg2 <- div_out %>% 
  filter(tool == 'Single Copy Genes') %>%
  ggplot(., aes(x = S, y = H, color = Agriculture)) +
  geom_jitter(size = 4) +
  scale_color_manual(values = colors[na.omit(unique(metadata$Agriculture))]) +
  theme_classic(base_size = 20) +
  theme(legend.position = 'none') +
  #labs(y = 'Entropy', x = 'Number of Genes')
  labs(y = '', x = '')




scg1 <- df %>%
    filter(tool == 'Single Copy Genes') %>%
  ggplot() +
  geom_point(aes(x = Outgroup, y = Escovopsis, size = NumberGenes)) +
  #geom_tile(aes(x = Outgroup, y = Escovopsis, fill = NumberGenes)) +
  geom_abline(intercept = 0, slope = 1, linetype = "dashed", colour = 'grey') +
  theme_classic(base_size = 20) +
   labs(y = 'Median Copy Number in Escovopsis',
         x = 'Median Copy Number in Outgroups') +
    labs(y = '',
        x = '') +
  scale_x_continuous(breaks = seq(0, max(df$Outgroup), by = 2),
                     labels = seq(0, max(df$Outgroup), by = 2)) +
  scale_y_continuous(breaks = seq(0, max(df$Escovopsis)),
                     labels = seq(0, max(df$Escovopsis))) +
  theme(legend.position = 'none') 

cowplot::plot_grid(plotlist = list(scg1, scg2), labels = 'AUTO') %>%
  ggsave2( filename = 'plots/single_copy_gene_diversity.png', height = 8, width = 12)



```



```{r, fig4A}

# x %>% 
#   data.frame() %>%
#     rownames_to_column('genome') %>%
#   gather(bgc, value, -genome) %>%
#   .$value %>% unique()



ant_ag <- c('Coral', 'Higher', 'Lower', 'Leafcutter')
x <- tools[1]
div_mats_ag <- lapply(tools, function(x){
  div_mat <- orthog_meta_ls %>% 
    mutate(Agriculture = ifelse(!Agriculture %in% ant_ag, yes = 'Outgroup', no = Agriculture)) %>%
    select(Agriculture, acc, tool, annotation) %>%
    filter(tool == x) %>%
    select(-tool) %>%
    mutate(count = 1) %>%
    group_by(acc, annotation) %>%
    mutate(n_annotation = sum(count)) %>%
    ungroup() %>%
    distinct() %>%
    group_by(Agriculture, annotation) %>%
    mutate(median_count = median(n_annotation)) %>%
    ungroup() %>%
    select(-acc, -count, -n_annotation) %>%
    distinct() %>%
    spread(Agriculture, median_count) %>%
    column_to_rownames(var = 'annotation')
  div_mat[is.na(div_mat)] <- 0
  div_mat <- div_mat[rowSums(div_mat) > 0, ]
  div_mat
})
names(div_mats_ag) <- tools


esco_genomes <- metadata %>% filter(!Agriculture %in% c('Trichoderma', "Hypomyces_Cladobotryum")) %>% .$acc

x <- div_mats[["BGC"]][esco_genomes,] %>% t()

x <- x[rowSums(x) >0,colSums(x)>0]


#colnames(x) = sub(",", "_", colnames(x))


pca_out    <- prcomp(x)
eigs       <- pca_out$sdev^2
proportion <- (eigs/sum(eigs))*100
cumulative <- cumsum(eigs)/sum(eigs)
n_k        <- 2




# data2 <- mds$points %>% 
#   data.frame() %>% 
#   rownames_to_column(var = 'acc') %>%
#   left_join(., metadata, by = 'acc')

data2 <- pca_out$rotation %>%
  data.frame() %>%
  rownames_to_column(var = 'acc') %>%
  left_join(., metadata, by = 'acc')

p <- ggplot(data2, aes(x = PC1, y = PC2)) + 
  geom_point(shape = 21, size = 3, aes(fill = Agriculture), 
             stroke = 0.5, colour = 'black') +
  theme_classic() +
  geom_text(data = data2, 
            aes(label = genus_species, color = NULL), nudge_x = -0.08, hjust = 0) + 
  scale_fill_manual(values = colors) +
  theme(legend.position = 'none')

p

p +
  ggsave('plots/figure4a.png', width = 7, height = 5)


p +
  ggsave('plots/figure4a.pdf', width = 7, height = 5)





m <- metadata %>% 
  select(Agriculture, acc) %>%  
  mutate(Agriculture = replace_na(Agriculture, 'Outgroup')) %>%
  distinct() %>% 
  column_to_rownames('acc')

annotation_row = data.frame(
     "Agriculture" = m[rownames(x), ],
     row.names = rownames(x),
    stringsAsFactors = FALSE
   )

annotation_col <- all_bgc %>%
  select(-acc, -n_genes_genome, -clade_groups, -Agriculture, -tool, -Orthogroup) %>%
  filter(annotation %in% colnames(x)) %>%
  distinct() %>%
  column_to_rownames('annotation')
  data.frame() 

ann_colors = list(
  'BGC_type' = c(NRPS = '#1B9E77', PKSI = '#D95F02', Terpene = '#7570B3', Others = "#E7298A", 'PKS-NRP' = "#66A61E"),
  'Agriculture' = colors[-which(names(colors) %in% c('Trichoderma', 'Hypomyces_Cladobotryum'))])


pheatmap::pheatmap(x,
                   legend = FALSE,
                   color = c('grey87', 'black'),
                   cluster_rows = TRUE,
                   cluster_cols = TRUE,
                   border_color = "grey70",
                   cellwidth = 10,
                   cellheight = 10,
                   annotation_row = annotation_row,
#                   annotation_col = annotation_col,
                   annotation_colors = ann_colors,
                   filename = paste0('plots/all_BGC_components_heatmap.pdf'))



pheatmap::pheatmap(x,
                   legend = FALSE,
                   color = c('grey87', 'black'),
                   cluster_rows = TRUE,
                   cluster_cols = TRUE,
                   border_color = "grey70",
                   cellwidth = 10,
                   cellheight = 10,
                   annotation_row = annotation_row,
#                   annotation_col = annotation_col,
                   annotation_colors = ann_colors,
                   filename = paste0('plots/all_BGC_components_heatmap.png'))

# library(ggord)
# 
# 
# 
# p <- ggord::ggord(mds, size = 2, alpha = 0.5) + 
#   theme_classic() 
# 
# p
# 


```

