---
title: "Genome Metadata and GFF3 Analysis"
author: "Kirsten Gotting"
date: "`r Sys.Date()`"
output:
  html_document:
    theme: yeti
    highlight: kate
    code_folding: show
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE, error = FALSE)
```

```{bash, eval = FALSE}
### stats files were made with agat with the following command
for i in `ls *gff`
do
  agat_sp_statistics.pl -o ${i%.gff}.stats -gff $i
  cat ${i%.gff}.stats | sed -e"s/ /_/g" | sed -e "s/___*/ /" >${i%.gff}.stats.fix
done
```



```{r}

library(tidyverse)
library(cowplot)
#library(ggpval)
library("googlesheets4")
options(scipen=999)


source('scripts/color_palettes.R')

metadata <- read_sheet("https://docs.google.com/spreadsheets/d/1IUjrth2EOe7UptgUs7lzKCVQk8c79-T5KmAg4tESLHY/edit#gid=0") %>%
  mutate(across(.cols = c('assembly_length', 'kmer_repeat_length_assemblies_bp', 'kmer_repeat_length_reads_bp'), .fns = function(x){x/1000000}))


#metadata <- read_csv('tables/metadata.csv') %>% data.frame() 

gff_f <- list.files('tables/gff_stats', full.names = TRUE, pattern = 'fix')

gff_stats <- lapply(gff_f, function(x){
  read_delim(x, delim = " ", col_names = c('stat','value')) %>% 
    mutate(acc = sub("\\..*$", "", basename(x)))
}) %>% bind_rows() %>% 
  data.frame(., stringsAsFactors = FALSE) 


gff_stats$acc <- unlist(lapply(gff_stats$acc, function(x){
  metadata[grep(x, metadata$acc), 'acc']
}))


unique(metadata$acc[!metadata$acc %in% gff_stats$acc])



```



```{r assembly_len}
data <- metadata %>% filter(!is.na(assembly_length)) %>% select(clade_groups, assembly_length) %>%
  filter(clade_groups %in% c('Escovopsis', 'Trichoderma'))





as_len <- metadata %>% filter(!is.na(assembly_length)) %>%
  ggplot(aes(x = clade_groups, y = assembly_length, colour = clade_groups)) + 
  geom_boxplot() + 
  geom_jitter(alpha = 0.5) +
  theme_classic() +
  labs(x = 'Groups', 
       y = 'Assembly Length (mb)')  +
  scale_y_continuous(limits = c(0, max(metadata$assembly_length))) +
  theme(
    legend.position = 'none') +
  scale_colour_manual(values = clade_colors) 




as_len
```


```{r rep_len_assembly}
rep_len <- metadata %>% 
  filter(!is.na(kmer_repeat_length_assemblies_bp)) %>%
  ggplot(aes(x = clade_groups, y = kmer_repeat_length_assemblies_bp, colour = clade_groups)) + 
  geom_boxplot() + 
  geom_jitter(alpha = 0.5) +
  theme_classic() +
  labs(y = 'Assembly Repeat Length (mb)', x = 'Groups')  +
  scale_y_continuous(limits = c(0, max(metadata$kmer_repeat_length_assemblies_bp))) +
  theme(
    legend.position = 'none') +
  scale_colour_manual(values = clade_colors) 

rep_len

```

```{r rep_len}
rep_len2 <- metadata %>% filter(!is.na(kmer_repeat_length_reads_bp)) %>%
  ggplot(aes(x = clade_groups, y = kmer_repeat_length_reads_bp, colour = clade_groups)) + 
  geom_boxplot() + 
  geom_jitter(alpha = 0.5) +
  theme_classic() +
  labs(y = 'k-mer Estimated Repeat Length (mb)', x = 'Groups')  +
  scale_y_continuous(limits = c(0, max(metadata$kmer_repeat_length_reads_bp))) +
  theme(
    legend.position = 'none') +
  scale_colour_manual(values = clade_colors) 

rep_len2

```


```{r genome_size}

genome_size <- metadata  %>%
  mutate(GenomeSize = as.numeric(GenomeSize)) %>%
  unnest(GenomeSize) %>%
  filter(!is.na(GenomeSize)) %>%
  ggplot(aes(x = clade_groups, 
             y = GenomeSize, 
             colour = clade_groups)) + 
  geom_boxplot() + 
  geom_jitter(alpha = 0.5) +
  theme_classic() +
  scale_colour_manual(values = clade_colors) +
  labs(x = 'Groups', 
       y = 'Genome Size (mb)') +
  scale_y_continuous(limits = c(0, 45))

genome_size

legend <- get_legend(
  # create some space to the left of the legend
  genome_size + theme(legend.box.margin = margin(0, 0, 0, 12))
)
genome_size <- genome_size + theme(legend.position = 'none')

```


```{r tables-mean-a-c}
meta_out <- metadata %>% 
  filter(!is.na(n_proteins), !is.na(assembly_length), !is.na(GenomeSize)) %>%
  select(clade_groups, n_proteins, assembly_length, GenomeSize) %>%
  gather(stat, value, -clade_groups) %>%
  group_by(stat, clade_groups) %>%
  summarize(mean = mean(value)) 

meta_out %>%
  knitr::kable(.,)
```




```{r mrna_len}



stats_interest_all <- c("Total_cds_length",
                        "mean_cds_length", 
                        'Number_of_intron_in_cds')

gff_stats <- gff_stats %>% 
  left_join(., metadata, by = 'acc') %>%
  mutate(across(.cols = c("Total_cds_length", "mean_cds_length", 'Number_of_intron_in_cds'), .fns = function(x){as.numeric(x)/1000000}))



gff_stats %>% select(stat, value, acc, genus) %>% filter(stat %in% stats_interest_all) %>% distinct() %>%  group_by(genus, acc, stat) %>%
  summarize(value = mean(value)) %>%
  ungroup() %>%
  spread(key = stat, value = value) %>%
  write_csv(., path = 'tables/supplementary_table_1-1.csv')




stats_interest <- c('Total_mrna_length')
sub <- gff_stats %>% filter(stat %in% stats_interest)
sub$stat <- factor(sub$stat, levels = stats_interest)

mrna_length<- sub %>% filter(stat %in% stats_interest) %>%
  ggplot(aes(x = clade_groups, y = value, colour = clade_groups)) + 
  geom_boxplot() + 
  geom_jitter(alpha = 0.5) +
  theme_classic() +
  ylab('Total mRNA Length (bp)') +
  xlab('Groups')+
  theme(
    legend.position = 'none') +
  scale_colour_manual(values = clade_colors) 

mrna_length

```

```{r num_introns}
stats_interest <- c('Number_of_intron_in_cds'
                    )

sub <- gff_stats %>% filter(stat %in% stats_interest)
sub$stat <- factor(sub$stat, levels = stats_interest)

number_introns <- ggplot(sub, aes(x = clade_groups, y = value, colour = clade_groups)) + 
  geom_boxplot() + 
  geom_jitter(alpha = 0.5) +
  theme_classic() +
  ylab('Number of Introns in CDS') +
  xlab('Groups')+
  theme(
    legend.position = 'none') +
  scale_colour_manual(values = clade_colors)
number_introns
```



```{r, eval = FALSE}
stats_interest <- c("Total_intron_length_per_cds")
sub <- gff_stats %>% filter(stat %in% stats_interest)
sub$stat <- factor(sub$stat, levels = stats_interest)

intron_length <- ggplot(sub, aes(x = clade_groups, y = value, colour = clade_groups)) + 
  geom_boxplot() + 
  geom_jitter(alpha = 0.5) +
  theme_classic() +
  ylab('Total Intron Length/CDS (bp)') +
  xlab('Groups')+
  theme(
    legend.position = 'none') +
  scale_colour_manual(values = clade_colors)
intron_length
##source('scripts/make_map.R')
##subplot(map, proteins, proteins, proteins, nrows = 2) 
```




```{r mean_cds}
stats_interest <- c("mean_cds_length" )
sub <- gff_stats %>% filter(stat %in% stats_interest)
sub$stat <- factor(sub$stat, levels = stats_interest)

mean_cds <- ggplot(sub, aes(x = clade_groups, y = value, colour = clade_groups)) + 
  geom_boxplot() + 
  geom_jitter(alpha = 0.5) +
  theme_classic() +
  ylab('Mean CDS Length (bp)') +
  xlab('Groups')+
  theme(
    legend.position = 'none') +
  scale_colour_manual(values = clade_colors)
mean_cds
##source('scripts/make_map.R')
##subplot(map, proteins, proteins, proteins, nrows = 2) 
```


```{r total_cds}
stats_interest <- c("Total_cds_length")
sub <- gff_stats %>% filter(stat %in% stats_interest) %>%
  mutate(value = value/1000000)
sub$stat <- factor(sub$stat, levels = stats_interest)

total_cds <- ggplot(sub, aes(x = clade_groups, y = value, colour = clade_groups)) + 
  geom_boxplot() + 
  geom_jitter(alpha = 0.5) +
  theme_classic() +
  ylab('Total CDS Length (mb)') +
  xlab('Groups')+
  theme(
    legend.position = 'none') +
  scale_colour_manual(values = clade_colors)
total_cds
##source('scripts/make_map.R')
##subplot(map, proteins, proteins, proteins, nrows = 2) 
```




```{r}
plot_grid(genome_size, as_len, total_cds, mean_cds, number_introns, rep_len, rep_len2, legend, labels = 'AUTO', label_size = 14,
          nrow = 4, ncol = 2) + 
  ggsave2(filename = 'plots/supplementary_fig2.pdf', height = 10, width = 10)

plot_grid(genome_size, as_len, total_cds, mean_cds, number_introns, rep_len, rep_len2, legend, labels = 'AUTO', label_size = 14,
          nrow = 4, ncol = 2) + 
  ggsave2(filename = 'plots/supplementary_fig2.png', height = 10, width = 10)
```

```{r}

stats_interest <- c("Total_cds_length", 'Number_of_intron_in_cds', 'mean_cds_length')
gff_out <- gff_stats %>% filter(stat %in% stats_interest) %>%
  select(clade_groups, stat,  value) %>%
  group_by(stat, clade_groups) %>%
  summarize(mean = mean(value),
            min = min(value),
            max = max(value)) 

gff_out %>%
  knitr::kable(.,)

```

```{r}

summ <- bind_rows(meta_out, gff_out)


## percent difference from escovopsis


lapply(unique(summ$stat), function(x){
    df <- summ %>% filter(stat == x)
    esco_val <- df %>% filter(clade_groups == 'Escovopsis') %>% .$mean
    df %>% mutate(perc_diff_escovopsis = (1- esco_val/mean)*100)
}) %>% bind_rows() %>%
  mutate(perc_diff_escovopsis = ifelse(clade_groups == 'Escovopsis', yes = NA, no = perc_diff_escovopsis)) %>%
  write_csv(path = 'tables/supplementary_table1.csv') 
 


```



```{r core-accessory}
load('rdata/orthologues_long.rdata')

ortho_full  <- orthog_long %>% 
  filter(!is.na(genes)) %>%
  group_by(genome) %>%
  mutate(acc = case_when(
      grepl('GCA_004303015', genome) ~ 'GCA_004303015.1',
      grepl('GCA_011799845', genome) ~ 'GCA_011799845.1',
      grepl('GCA', genome) ~ paste(strsplit(genome, split = "_")[[1]][c(1,2)], collapse = "_"),
      grepl("SPDT00000000.1_genomic", genome) ~ "SPDT00000000",
      grepl('all.maker', genome) ~ sub(".all.maker.proteins", '', genome),
      grepl('LGSR', genome) ~ "LGSR00000000",
      TRUE ~ genome
    )) %>%
  ungroup() %>%
  dplyr::rename('gene' = genes) %>%
  left_join(., metadata, by = 'acc')

core <- orthog_long %>%
  group_by(Orthogroup) %>%
  filter(!is.na(genes), genome != 'GCA_000225605.1_CmilitarisCM01_v01_protein') %>%
  mutate(n_genomes = length(unique(genome))) %>%
  filter(n_genomes == 45)

## 1) genome-specific genes are those thatdo not cluster with any gene in any other genome in the phylogeny 
acces1 <- orthog_long %>%
  group_by(Orthogroup) %>%
  filter(!is.na(genes), genome != 'GCA_000225605.1_CmilitarisCM01_v01_protein') %>%
  mutate(n_genomes = length(unique(genome))) %>%
  filter(n_genomes == 1)

# 2) lineage-specific genes are those that have homologs in only the evolutionarilyclosest group of genomes 
acces2 <- ortho_full %>%
  filter(!is.na(gene), genome != 'GCA_000225605.1_CmilitarisCM01_v01_protein') %>%
  group_by(Orthogroup) %>%
  mutate(n_ags = length(unique(Agriculture))) %>%
  filter(n_ags == 1)
  
```