---
title: "Genome Metadata and GFF3 Analysis"
author: "Kirsten Gotting"
date: "`r Sys.Date()`"
output:
  html_document:
    theme: yeti
    highlight: kate
    code_folding: show
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE, error = FALSE)
```



```{r}

library(tidyverse)
library(cowplot)
options(scipen=999)



source('scripts/color_palettes.R')
metadata <- read_csv('tables/metadata.csv') %>% data.frame()

proteins <- dplyr::filter(metadata, !is.na(n_proteins)) %>%
  ggplot(aes(x = clade_groups, y = n_proteins, colour = clade_groups)) + 
  geom_boxplot() + 
  geom_jitter(alpha = 0.5) +
  theme_classic() +
  labs(x = '', 
       y = 'Number of Genes')  +
  scale_y_continuous(limits = c(0, max(metadata$n_proteins))) +
  theme(
        legend.position = 'none',
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank()) +
  scale_colour_manual(values = clade_colors) + 
  ggsave('plots/protein_coding_gene_N_boxplot.pdf', width = 3, height = 3)
proteins


```



```{r}
as_len <- metadata %>% filter(!is.na(assembly_length)) %>%
  ggplot(aes(x = clade_groups, y = assembly_length, colour = clade_groups)) + 
  geom_boxplot() + 
  geom_jitter(alpha = 0.5) +
  theme_classic() +
  labs(x = '', 
       y = 'Assembly Length (bp)')  +
  scale_y_continuous(limits = c(0, max(metadata$assembly_length))) +
  theme(
    legend.position = 'none',
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank()) +
  scale_colour_manual(values = clade_colors) 

as_len
```

```{r}
gs_shape_data <- filter(metadata, 
                        GenomeSizeEstimationStrategy %in% c('FI-PC', 'pacbio'))
genome_size <- 
  metadata %>% filter(!is.na(GenomeSize)) %>%
  ggplot(aes(x = clade_groups, 
             y = GenomeSize, 
             colour = clade_groups)) + 
  geom_boxplot() + 
  geom_jitter(alpha = 0.5, aes(shape = GenomeSizeEstimationStrategy), size = 3) +
  theme_classic() +
  scale_colour_manual(values = clade_colors) +
  # geom_jitter(data = gs_shape_data, aes(shape = GenomeSizeEstimationStrategy),
  #             colour = 'black', width = 0.25) +
  labs(x = '', 
       y = 'Genome Size (mb)') +
  scale_y_continuous(limits = c(0, max(metadata$GenomeSize))) +
  theme(
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank())
genome_size

legend <- get_legend(
  # create some space to the left of the legend
  genome_size + theme(legend.box.margin = margin(0, 0, 0, 12))
)
genome_size <- genome_size + theme(legend.position = 'none')

```


```{r tables-mean-a-c}
meta_out <- metadata %>% 
  filter(!is.na(n_proteins), !is.na(assembly_length), !is.na(GenomeSize)) %>%
  select(clade_groups, n_proteins, assembly_length, GenomeSize) %>%
  gather(stat, value, -clade_groups) %>%
  group_by(stat, clade_groups) %>%
  summarize(mean = mean(value)) 

meta_out %>%
  knitr::kable(.,)
```


```{bash, eval = FALSE}
### stats files were made with agat with the following command
for i in `ls *gff`
do
  agat_sp_statistics.pl -o ${i%.gff}.stats -gff $i
  cat ${i%.gff}.stats | sed -e"s/ /_/g" | sed -e "s/___*/ /" >${i%.gff}.stats.fix
done
```



```{r}
gff_f <- list.files('gff_stats', full.names = TRUE, pattern = 'fix')
gff_stats <- lapply(gff_f, function(x){
  read_delim(x, delim = " ", col_names = c('stat','value')) %>% 
    mutate(filename = sub("_genomic", '', sub(".stats.fix", "", basename(x))))
}) %>% bind_rows() %>% data.frame(., stringsAsFactors = FALSE)

gff_stats$acc <- unlist(lapply(gff_stats$filename, function(x){
  metadata[grep(x, metadata$genome_id), 'acc']
}))


gff_stats <- gff_stats %>% left_join(., metadata, by = 'acc')

stats_interest <- c('Total_mrna_length')
sub <- gff_stats %>% filter(stat %in% stats_interest)
sub$stat <- factor(sub$stat, levels = stats_interest)

mrna_length<- sub %>% filter(stat %in% stats_interest) %>%
  ggplot(aes(x = clade_groups, y = value, colour = clade_groups)) + 
  geom_boxplot() + 
  geom_jitter(alpha = 0.5) +
  theme_classic() +
  ylab('Total mRNA Length (bp)') +
  xlab('')+
  theme(
    legend.position = 'none',
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank()) +
  scale_colour_manual(values = clade_colors) 

mrna_length

```

```{r}
stats_interest <- c('Number_of_intron_in_cds'
                    )

sub <- gff_stats %>% filter(stat %in% stats_interest)
sub$stat <- factor(sub$stat, levels = stats_interest)

number_introns <- ggplot(sub, aes(x = clade_groups, y = value, colour = clade_groups)) + 
  geom_boxplot() + 
  geom_jitter(alpha = 0.5) +
  theme_classic() +
  ylab('Number of Introns in CDS') +
  xlab('')+
  theme(
    legend.position = 'none',
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank()) +
  scale_colour_manual(values = clade_colors)
number_introns
```



```{r}
stats_interest <- c("Total_intron_length_per_cds")
sub <- gff_stats %>% filter(stat %in% stats_interest)
sub$stat <- factor(sub$stat, levels = stats_interest)

intron_length <- ggplot(sub, aes(x = clade_groups, y = value, colour = clade_groups)) + 
  geom_boxplot() + 
  geom_jitter(alpha = 0.5) +
  theme_classic() +
  ylab('Total Intron Length/CDS (bp)') +
  xlab('')+
  theme(
    legend.position = 'none',
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank()) +
  scale_colour_manual(values = clade_colors)
intron_length
##source('scripts/make_map.R')
##subplot(map, proteins, proteins, proteins, nrows = 2) 
```




```{r}
stats_interest <- c("mean_mrna_length")
sub <- gff_stats %>% filter(stat %in% stats_interest)
sub$stat <- factor(sub$stat, levels = stats_interest)

mean_mrnas <- ggplot(sub, aes(x = clade_groups, y = value, colour = clade_groups)) + 
  geom_boxplot() + 
  geom_jitter(alpha = 0.5) +
  theme_classic() +
  ylab('Mean mRNA Length (bp)') +
  xlab('')+
  theme(
    legend.position = 'none',
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank()) +
  scale_colour_manual(values = clade_colors)
mean_mrnas
##source('scripts/make_map.R')
##subplot(map, proteins, proteins, proteins, nrows = 2) 
```



```{r}
plot_grid(genome_size, as_len, mrna_length, mean_mrnas, number_introns, intron_length, legend, labels = 'AUTO', label_size = 14,
          nrow = 4, ncol = 2) + 
  ggsave2(filename = 'plots/supplementary_fig1.pdf', height = 10, width = 10)

```

```{r}

stats_interest <- c('Total_mrna_length', 'Number_of_intron_in_cds', 'Number_of_intron_in_cds', "Total_intron_length_per_cds", 'mean_mrna_length')
gff_out <- gff_stats %>% filter(stat %in% stats_interest) %>%
  select(clade_groups, stat,  value) %>%
  group_by(stat, clade_groups) %>%
  summarize(mean = mean(value)) 

gff_out %>%
  knitr::kable(.,)

```

```{r}

summ <- bind_rows(meta_out, gff_out)


## percent difference from escovopsis


lapply(unique(summ$stat), function(x){
    df <- summ %>% filter(stat == x)
    esco_val <- df %>% filter(clade_groups == 'Escovopsis') %>% .$mean
    df %>% mutate(perc_diff_escovopsis = (1- esco_val/mean)*100)
}) %>% bind_rows() %>%
  mutate(perc_diff_escovopsis = ifelse(clade_groups == 'Escovopsis', yes = NA, no = perc_diff_escovopsis)) %>%
  write_csv(path = 'tables/supplementary_table1.csv') 
 


```





