---
title: "Orthologous Genes"
author: "Kirsten Gotting"
date: "`r Sys.Date()`"
output:
  html_document:
    theme: yeti
    highlight: kate
    code_folding: show
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE, error = FALSE)
knitr::opts_knit$set(root.dir = '..')

```


```{r begin}
#library(KEGGREST)
library(GO.db)
#library(GSEABase)
library(pheatmap)
library(RColorBrewer)
library(VennDiagram)
library(UpSetR)
library(tidyverse)
library(cowplot)


source('scripts/color_palettes.R')

tools_interest <- c('CARD', 
                    'cazy', 
                    'led', 
                    'MEROPS', 
                    'virulence', 
                    'antismash',
                    'other')

```

```{r create_data_inputs}
annot <- read_tsv('annotation/all_annotations.txt', col_names = c('acc', 'gene', 'tool', 'annot'))

meta <- read_csv('tables/metadata.csv')

all_gene_id_files <- list.files('annotation/gene_ids/', full.names = TRUE)

all_gene_ids <- lapply(all_gene_id_files, function(x){
  ids <- scan(x, what = 'character')
  data.frame(gene = ids, genome = rep(basename(x), length(ids)), 
             stringsAsFactors = FALSE) %>%
    mutate(acc = case_when(
      grepl('GCA_004303015', genome) ~ 'GCA_004303015.1',
      grepl('GCA_011799845', genome) ~ 'GCA_011799845.1',
      grepl('GCA', genome) ~ paste(strsplit(genome, split = "_")[[1]][c(1,2)], collapse = "_"),
      grepl("SPDT00000000.1_genomic", genome) ~ "SPDT00000000",
      grepl('LGSR', genome) ~ "LGSR00000000",
      grepl('.fasta_gene_ids', genome) ~ sub(".all.maker.proteins.fasta_gene_ids", '', genome),
      TRUE ~ genome
    )) 
}) %>% bind_rows() %>%
  dplyr::select(-genome)



#orthog      <- read_tsv('tables/Orthogroups.tsv')


load('rdata/orthologues_long.rdata')



## join in the orthologues with singleton gene ids for each genome
ortho_full  <- orthog_long %>% 
  filter(!is.na(genes)) %>%
  group_by(genome) %>%
  mutate(acc = case_when(
      grepl('GCA_004303015', genome) ~ 'GCA_004303015.1',
      grepl('GCA_011799845', genome) ~ 'GCA_011799845.1',
      grepl('GCA', genome) ~ paste(strsplit(genome, split = "_")[[1]][c(1,2)], collapse = "_"),
      grepl("SPDT00000000.1_genomic", genome) ~ "SPDT00000000",
      grepl('all.maker', genome) ~ sub(".all.maker.proteins", '', genome),
      grepl('LGSR', genome) ~ "LGSR00000000",
      TRUE ~ genome
    )) %>%
  ungroup() %>%
  dplyr::rename('gene' = genes) %>%
  full_join(all_gene_ids, ., by = c('acc', 'gene')) %>%
  mutate(orthology_meta = ifelse(is.na(Orthogroup),
                                 yes = gene,
                                 no = Orthogroup)) %>%
  left_join(., meta, by = 'acc') %>%
  filter(!is.na(Agriculture)) %>%
  group_by(Agriculture, orthology_meta) %>%
  mutate(n_genomes = length(unique(acc))) %>%
  mutate(n_copies_per_genome = length(acc)) %>%
  ungroup()

```

```{r create_data_venns}

## compare the three pangenomes of the three clades

groups <- unique(meta$Agriculture) %>% na.omit()

ag <- c("Lower",
         "Coral",
         "Higher",
         "Leafcutter")


x <- groups[6]
venn1 <- lapply(groups, function(x){
  n_g =   ortho_full %>% filter(Agriculture %in% x) %>% .$genome_id %>% unique() %>% length()
  if(n_g > 4){
    cutoff <- trunc(n_g *.95)
  } else{
    cutoff <- n_g  # Orthogroup must be in all of the genomes if there are less than 5 representatives (Hypomyces clade)
  }
  ortho_full %>% 
    filter(Agriculture %in% x) %>%
    filter(n_genomes >= cutoff)
}) %>% 
  bind_rows() %>%
  dplyr::select(orthology_meta, clade_groups, Agriculture, n_genomes) %>%
  distinct() %>%
  group_by(clade_groups, Agriculture, orthology_meta) %>%
  summarize(genome_count = sum(n_genomes)) %>%
  ungroup()


venn1 %>%
  group_by(clade_groups) %>%
  summarize(mean(genome_count))

venn <- venn1
remove(venn1)




write_csv(venn, path = 'tables/clade_orthogroup_venn.csv')




listInput <- lapply(c(groups), 
                    function(x){
                      venn %>% filter(Agriculture == x) %>% .$orthology_meta %>% unique()
                    })

names(listInput) <- c(groups)
lapply(listInput, head)




listInput[['AllOutgroup']] <- sort(unique(c(listInput$Trichoderma, listInput$Hypomyces_Cladobotryum)))
listInput[['Escovopsis']] <- sort(unique(c(listInput$Coral, listInput$Higher, listInput$Leafcutter, listInput$Lower)))


# listInput <- list('Escovopsis' = esco_set,, 
#                   'Hypomyces/Cladobotryum' = ch_set,
#                   'Trichoderma' = t_set)
# set_name(listInput)

lapply(listInput, head)


input = listInput
elements <- unique(unlist(input))
data <- unlist(lapply(input, function(x) {
  x <- as.vector(match(elements, x))
}))
data[is.na(data)] <- as.integer(0)
data[data != 0] <- as.integer(1)
data <- data.frame(matrix(data, ncol = length(input), byrow = F))
data <- data[which(rowSums(data) != 0), ]
names(data) <- names(input)
rownames(data) <- elements

orthogroup_pa_ag <- data %>% 
  data.frame() %>%
  rownames_to_column('orthogroup') %>%
  gather(group, presence, -orthogroup) %>%
  filter(!group %in% c('AllOutgroup', 'Escovopsis')) %>%
  group_by(orthogroup) %>% 
  mutate(lineage_specific = ifelse(sum(presence) == 1, TRUE, FALSE)) %>%
  # mutate(lineage_specific = ifelse(presence == 1 & lineage_specific == TRUE, TRUE, FALSE)) %>%
  ungroup()

head(orthogroup_pa_ag)
  
#orthogroup_pa_ag %>% filter(lineage_specific == TRUE)
#orthogroup_pa_ag %>% filter(orthogroup == 'OG0000202')


# lapply(unique(orthogroup_pa_ag$group), function(x){
#   print(x)
#   print(filter(orthogroup_pa_ag, group == x, lineage_specific == TRUE))
# })


data <- NULL

save(orthogroup_pa_ag, file = 'rdata/orthogroup_pa_ag.rdata')



```

```{r}

unique(c(listInput$Hypomyces_Cladobotryum, listInput$Escovopsis, listInput$Trichoderma)) %>%
  write(., file = 'anvio_genes_include.txt')



```

```{r plot_upset}








pdf("plots/orthologues_upset_all.pdf", width = 15, height = 8)
upset(fromList(listInput), order.by = "freq", 
      sets = c('AllOutgroup', 'Lower','Coral','Higher','Leafcutter'),
       point.size = 3.5, 
       line.size = 1.5, 
       text.scale = c(1.3, 1.3, 1, 1, 2, 0.75),
       mb.ratio = c(0.8, 0.2))
dev.off()

#pdf("plots/orthologues_upset_subset.pdf", width = 8, height = 8)





pdf("plots/orthologues_upset_clade.pdf", width = 8, height = 5)
upset(fromList(listInput), order.by = "freq",
      sets = c('Hypomyces_Cladobotryum', 'Escovopsis', 'Trichoderma'),
       point.size = 3.5,
       line.size = 1.5,
       text.scale = c(1.3, 1.3, 1, 1, 2, 1),
       mb.ratio = c(0.8, 0.2))
dev.off()
# 
# 
#  ## Keep this as it show a combination of the set size of esco, with the intersection size with the others
# set_data <- data.frame(
#   'Hypomyces_Cladobotryum' = length(listInput$`Hypomyces_Cladobotryum`),
#   Trichoderma = length(listInput$Trichoderma),
#   Escovopsis = length(listInput$Escovopsis),
#            Overlap = length(union(which(listInput$Escovopsis %in% listInput$`Hypomyces_Cladobotryum`),
#                                       which(listInput$Escovopsis %in% listInput$Trichoderma)))) %>%
#   gather(set, n_genes)
# 
# set_data$set <- factor(set_data$set, levels =rev(c('Trichoderma', 'Hypomyces_Cladobotryum', 'Escovopsis', 'Overlap')))
# 
# ggplot(set_data, aes(y = n_genes, x = set, fill = set)) +
#   geom_col() +
#   theme_classic() +
#   scale_fill_manual(values = clade_colors) +
#   labs(y = '',
#        x = '') +
#   geom_text(aes(y = n_genes +80, label = n_genes)) +
#   scale_y_continuous(expand = c(0,0), limits = c(0, max(set_data$n_genes)+200)) +
#   theme(axis.text.x = element_blank(),
#         axis.ticks = element_blank(),
#         legend.justification=c(1,1),
#         ) +
#   ggsave('plots/esco_v_overlap_barchart.pdf', width = 4.5, height = 5)



set_data_all <- fromList(listInput)
colnames(set_data_all) <- names(listInput)

data <- set_data_all %>% 
  dplyr::select(-Hypomyces_Cladobotryum, -Trichoderma, -Escovopsis) %>%
  group_by_all() %>% 
  summarise(count = n()) %>%
  ungroup() %>%
  filter(!count %in% c(1507, 4988)) %>%  ## filtering out all matches and all outgroup
  mutate(rows = seq(1, nrow(.), by = 1)) %>%
  gather(group_name, presence, -count, -rows) %>%
  filter(presence != 0)

data$group_name <- factor(data$group_name, levels = rev(c('AllOutgroup', 'Lower', 'Coral', 'Higher', 'Leafcutter')))

# 
# taken = c()
# list_out <- list()
# levels_sub = rev(levels(data$group_name))
# 
# for (i in levels_sub) {
#   rows_s <- data %>% filter(group_name == i, group_name == y) %>% .$rows  
#   rows = rows[!rows %in% taken]
#   list_out[[i]] = rows
#   taken = c(taken, rows)
#   }


# df_join <- lapply(names(list_out), function(x){
#   data.frame(group_plot = x, rows = list_out[[x]])
# }) %>% bind_rows()


## remove? 
# data <- data %>% 
#   left_join(df_join, ., by = 'rows')  
# 
# data$rows <- factor(data$rows, levels = unique(data$rows))
# data$group_plot <- factor(data$group_plot, levels = unique(data$group_plot))





```


```{r plot_upset_subset}



fais <- list.files('annotation/proteins/', pattern = 'fai', full.names = TRUE)
aa_lengths <- lapply(fais, function(x){
  read_tsv(x, col_names = c('gene', 'length', NA, NA, NA)) %>%
    dplyr::select(-contains('X'))
}) %>% bind_rows()
# 
# 
groups1 = list(
  c('Lower','AllOutgroup'),
  c('Coral', 'AllOutgroup'),
  c('Higher', 'AllOutgroup'),
  c('Leafcutter', 'AllOutgroup'),
  c('Coral', 'Lower', 'AllOutgroup'),
  c('Coral', 'Higher', 'AllOutgroup'),
  c('Higher', 'Leafcutter', 'AllOutgroup'),  
  c('Higher', 'Lower', 'AllOutgroup'),
  c('Leafcutter', 'Lower', 'AllOutgroup'),
  c('Coral', 'Leafcutter', 'AllOutgroup'),
  c('Coral', 'Higher', 'Lower', 'AllOutgroup'),
  c('Coral', 'Leafcutter', 'Lower', 'AllOutgroup'),
  c('Coral', 'Higher', 'Leafcutter', 'AllOutgroup'),
  c('Higher', 'Leafcutter', 'Lower', 'AllOutgroup'),
  c('Lower'),  
  c("Coral", 'Lower'),  
  c('Lower', 'Higher'),
  c('Lower', 'Leafcutter'),  
  c('Coral', 'Lower', 'Higher'),
  c('Leafcutter', 'Lower', 'Higher'),
  c("Coral", 'Lower', 'Leafcutter'),
  c('Coral', 'Higher', 'Leafcutter', 'Lower'),
  c('Coral'),
  c('Coral', 'Higher'),
  c("Coral", 'Leafcutter'),
  c("Coral", 'Higher', 'Leafcutter'),
  c('Higher'),
  c('Higher', 'Leafcutter'),
  c('Leafcutter'))



groups_sub = list(
              c('Higher', 'Leafcutter', 'Lower', 'AllOutgroup'),
              c('Lower','AllOutgroup'),
              c('Higher', 'Leafcutter', 'AllOutgroup'),
              c('Coral', 'Higher', 'Lower', 'AllOutgroup'),
              c('Coral', 'Higher', 'Leafcutter', 'AllOutgroup'),
              c('Coral', 'Leafcutter', 'Lower', 'AllOutgroup'),
              c('Coral', 'Lower', 'AllOutgroup'),
              c('Higher', 'Leafcutter', 'AllOutgroup'),
              c('Leafcutter'),
              c('Lower'),
              c('Higher', 'Leafcutter'),
              c('Higher', 'Lower', 'AllOutgroup'),
              c('Leafcutter', 'Lower', 'AllOutgroup'),
              c('Leafcutter', 'AllOutgroup'),
              c('Higher', 'AllOutgroup'),
              c('Lower'),
              c('Coral', 'AllOutgroup'),
              c('Lower', 'Leafcutter', 'Higher'),
              c('Coral', 'Higher', 'AllOutgroup'),
              c('Coral', 'Leafcutter', 'AllOutgroup')) %>% 
  unique()


groups <- lapply(groups1, function(x){
  paste(sort(x), collapse = ",")
}) %>% unlist()



input = listInput[c(ag, 'AllOutgroup')]
elements <- unique(unlist(input))


data <- unlist(lapply(input, function(x) {
  x <- as.vector(match(elements, x))
}))

data[is.na(data)] <- as.integer(0)
data[data != 0] <- as.integer(1)
data <- data.frame(matrix(data, ncol = length(input), byrow = F))
data <- data[which(rowSums(data) != 0), ]
names(data) <- paste0("pa_", names(input))
rownames(data) <- elements
data$Orthogroup <- elements




annot_add <- annot %>%
   left_join(., ortho_full, by = 'gene') %>%
   left_join(., aa_lengths, by = 'gene') %>%
   left_join(data, ., by = c('Orthogroup')) %>%
  filter(annot != '-', 
         annot != 'N') %>%
  distinct()

counts <- c()
data <- data[, paste0("pa_", c(ag, 'AllOutgroup'))]
cols <- colnames(data[, paste0('pa_', c(ag, 'AllOutgroup'))])


for (x in rownames(data)) {
  id = sort(cols[which(data[x, ] == 1)])
  id = paste(id, collapse = ",") %>% gsub("pa_", "", .)
  if(!id %in% names(counts)){
    counts[id] = x
  }  else {
    counts[id] = paste(counts[id], x, sep = ",")
  }
}

counts <- counts[groups]

counts_df <- lapply(names(counts), function(x){
  genes = strsplit(counts[x], split = ',')[[1]]
  data.frame(group = rep(x, length(genes)), Orthogroup = genes)
}) %>% bind_rows()




orthogroup_meta <- annot_add %>% 
  group_by(Orthogroup, tool) %>%
  mutate(annot = strsplit(annot, split = ",")) %>%
  unnest(annot) %>%
  mutate(combined_annot = paste(unique(annot), collapse = ",")) %>%
  ungroup()  %>%
  group_by(Orthogroup) %>%
  mutate(average_length = mean(length)) %>%
  ungroup() %>%
  dplyr::select(Orthogroup,
         tool, 
         combined_annot,
         contains('pa_'),
         average_length) %>%
  distinct() %>%
  left_join(., counts_df, by = 'Orthogroup') %>%
  distinct()



ortho_subset <- orthogroup_meta %>% 
  filter(!is.na(group)) %>%
  group_by(group, tool) %>%
  mutate(tool_keep = case_when(
    tool %in% c(tools_interest) ~ tool,
    TRUE ~ 'other'
  )) %>%
  ungroup() %>%
  dplyr::select(Orthogroup, tool_keep, group) %>%
  distinct() %>%
  mutate(annot_pa = 1)   %>%
  group_by(Orthogroup) %>%
  mutate(num_annots = sum(annot_pa)) %>%
  ungroup() %>%
  group_by(Orthogroup, tool_keep) %>%
  mutate(annot_pa = ifelse(num_annots > 1 & tool_keep == 'other',
                           yes = 0, 
                           no = 1)) %>%
  ungroup() %>%
  dplyr::select(-num_annots) %>%
  spread(tool_keep, annot_pa) %>%
  dplyr::rename('BGC' = antismash,
         'Resistance' = CARD,
         'CAZyme' = cazy,
         'Lipase' = led,
         'Peptidase' = MEROPS,
         'Virulence' = virulence)
  

ortho_subset[is.na(ortho_subset)] <- 0
  
ortho_subset %>%
  mutate(group = list(strsplit(group, split = ',')[[1]])) %>%
  unnest_longer(group) %>%
  write_csv('tables/orthogroup_group_functional_annotation.csv')


pdf("plots/orthologues_upset_subset.pdf", width = 8, height = 8)

p <- upset(fromList(listInput), 
      order.by = c("freq"), 
      sets = c('AllOutgroup', 'Lower','Coral','Higher','Leafcutter'),
      intersections = groups1,
      point.size = 3.5, 
      line.size = 1.5, 
      text.scale = c(1.3, 1.3, 1, 1, 2, 0.75),
      mb.ratio = c(0.8, 0.2))

dev.off()


# pdf("plots/orthologues_upset_full.pdf", width = 8, height = 8)
# upset(fromList(listInput), 
#       order.by = c("freq"), 
#       sets = c('AllOutgroup', 'Lower','Coral','Higher','Leafcutter'),
#       point.size = 3.5, 
#       line.size = 1.5, 
#       text.scale = c(1.3, 1.3, 1, 1, 2, 0.75),
#       mb.ratio = c(0.8, 0.2))
# dev.off()


## fix 

data$rowsums <- rowSums(data)

list_out_in_upset = data %>% 
  filter(!(pa_AllOutgroup == 1 & rowsums == 1), rowsums != 5) %>% 
  rownames(.)


write(list_out_in_upset, file = 'rdata/anvio_ortho.txt', sep = "\n")

# legend_p <- ortho_subset %>% 
#     gather(annot, count, -Orthogroup, -group) %>%
#     filter(group == 'AllOutgroup,Higher,Leafcutter,Lower') %>%
#     group_by(group, annot) %>%
#     summarize(count_sum = sum(count)) %>% 
#     group_by(group) %>%
#     mutate(total_count = sum(count_sum)) %>%
#     group_by(group, annot) %>%
#     summarize(proportion = unique((count_sum/total_count)*100)) %>%
#     ggplot(aes(x = group, y = proportion, fill = annot)) +
#     geom_col() + 
#     scale_fill_manual(values = annot_colors) +
#     theme_classic()
# 
# 
# legend <- get_legend(
#   # create some space to the left of the legend
#   legend_p + theme(legend.box.margin = margin(0, 0, 0, 12))
# )


```



```{r groupEnrichment, eval = FALSE}
library(topGO)
## nothing gets enriched here
go_annot <- annot_add %>% filter(tool == 'GO') %>%  
  dplyr::select(Orthogroup,
         annot,
         contains('pa_'), Agriculture) %>% 
  group_by(Orthogroup) %>%
  mutate(AllAgs = paste(sort(unique(Agriculture)), collapse = ',')) %>%
  ungroup() %>%
  dplyr::select(-Agriculture) %>%
  distinct() %>%
  gather(agric, val, -Orthogroup, -annot, -AllAgs) %>%
  dplyr::select(-agric) %>%
  group_by(Orthogroup) %>%
  mutate(go = paste(unique(strsplit(annot, ",")[[1]]), collapse = ",")) %>%
  ungroup() %>%
  dplyr::select(-annot) %>%
  distinct()

go_annot_u      <- go_annot %>% dplyr::select(go, Orthogroup) %>% distinct()
ortho2go        <- go_annot_u$go
names(ortho2go) <- go_annot_u$Orthogroup


ortho2go        <- lapply(ortho2go, function(x){strsplit(x, ',')[[1]]})
geneNames        <- names(ortho2go)


groups <- lapply(groups, function(x){
  paste(sort(x), collapse = ",")
}) %>% unlist()


colnames(data) <- sub('pa_', '', colnames(data))
counts <- c()
data <- data[, c(ag, 'AllOutgroup')]
cols <- colnames(data[, c(ag, 'AllOutgroup')]) 


for (x in rownames(data)) {
  id = sort(cols[which(data[x, ] == 1)])
  id = paste(id, collapse = ",")
  if(!id %in% names(counts)){
    counts[id] = x
  }  else {
    counts[id] = paste(counts[id], x, sep = ",")
  }
}

xx <- as.list(GOTERM)

library(topGO)
go_enrich <- lapply(groups, function(x){
  ## define the genes that I'm interested in (those not in fungus farming ant escovopsis)
  myInterestingGenes <- strsplit(counts[x], split = ",")[[1]]
  geneList           <- factor(as.integer(geneNames %in% myInterestingGenes))
  names(geneList)    <- geneNames
  
  
  ## perform an enrichment test on the genes that are not in FFA
  enrich <- lapply(c('BP', 'MF', 'CC'), function(x){
    GOdata <- new("topGOdata", ontology = x, allGenes = geneList,
                  annot = annFUN.gene2GO, gene2GO = ortho2go)
    test.stat <- new("classicCount", 
                     testStatistic = GOFisherTest, 
                     name = "Fisher test")
    resultFisher         <- getSigGroups(GOdata, test.stat)
    pVal                 <- data.frame(pval=signif(score(resultFisher), 6),BH.correction=seq(1, length(score(resultFisher))))
    pVal                 <- pVal[order(pVal$pval),]
    pVal$BH.correction   <- signif(p.adjust(pVal$pval, method="BH"), 6)
    pVal.sub             <- pVal[pVal$BH.correction<0.01,]
    pVal.sub$go_id       <- rownames(pVal.sub)
    pVal.try <- cbind(pVal.sub, Term=sapply(pVal.sub$go_id, FUN=function(n){Term(xx[[n]])}), Ontol=sapply(pVal.sub$go_id, FUN=function(n){Ontology(xx[[n]])}))
    
    if (nrow(pVal.try) >0){
      df2                  <- pVal.try[,c("go_id", "Term", "Ontol", "pval", "BH.correction")]
      out <- list(df2, GO, resultFisher)
      invisible(out)
    } else {
      message("No enrichment")
    }
  })
})


  
### KEGG enrichment test



# path <- keggList("pathway")
# head(path)
# 
# ko_to_path <- keggLink("ko", 'pathway')
# ko_path_df <- data.frame(KO = ko_to_path, path_id = names(ko_to_path))
# 
# 
# 
# keggframeData = distinct(data.frame(kegg = sub('ko:K', '', annot$KO), gene = annot$OrthoGroup, 
#                          stringsAsFactors = FALSE))
# head(keggframeData)
# keggFrame=KEGGFrame(keggframeData, organism = 'Escovopsis')
# 
# gsc <- GeneSetCollection(keggFrame, setType = KEGGCollection())
# 
# universe = unique(keggframeData$gene)
# 
# 
# genes = orthogroup_meta %>% filter(group == 'Leafcutter') %>% .$OrthoGroup
# 
# kparams <- GSEAKEGGHyperGParams(name="My Custom GSEA based annot Params",
#                                   geneSetCollection=gsc,
#                                   geneIds = genes,
#                                   universeGeneIds = universe,
#                                   pvalueCutoff = 0.05,
#                                   testDirection = "over")
# kOver <- hyperGTest(kparams)
# 
# results <- summary(kOver) %>%
#   data.frame() %>%
#   mutate(BH.correction = signif(p.adjust(Pvalue,method="BH"), 6))  %>%
#   mutate(KO = paste0('ko:K', KEGGID)) %>%
#   left_join(., ko_path_df, 'KO') %>%
#   mutate(path_def = path[path_id]) %>%
#   filter(!is.na(path_def),
#          BH.correction <= 0.05)
#    
# head(arrange(results, Pvalue), 50) %>% .$path_def

# venn.diagram(
#   x = list('Escovopsis' = esco_set, 'Hypomyces/Cladobotryum' = ch_set, 'Trichoderma' = t_set),
#   height = 3000,
#   width = 3000,
#    category.names = c(paste0("Escovopsis\n(", n_distinct(esco_set), " Orthologues across ", nrow(filter(meta, clade_groups == 'Escovopsis')), " genomes)"),
#                       paste0("Cladobotryum_Hypomyces\n(", n_distinct(ch_set), " Orthogroups across ", nrow(filter(meta, clade_groups == 'Hypomyces/Cladobotryum')), " genomes)"), 
#                       paste0("Trichoderma\n(", n_distinct(t_set), " Orthogroups across ", nrow(filter(meta, clade_groups == 'Trichoderma')), " genomes)")),
#   filename = 'plots/shared_genes_venn_diagramm.png',
#   resolution = 300,
#   # Numbers
#   fontface = "bold",
#   fontfamily = "sans",
#   main.fontface = 'bold',
#   main.fontfamily = 'sans',
#   # Set names
#   cat.fontface = "bold",
#   cat.default.pos = "outer",
#   cat.pos = c(-27, 27, 135),
#   cat.dist = c(0.055, 0.055, 0.085),
#   cat.fontfamily = "sans",
#   rotation = 1
# )



# ortho_annot <- ortho_full %>%
#   mutate(cazy = ifelse(gene %in% cazy,
#                        yes = 'cazy',
#                        no = NA)) %>%
#   mutate(card = ifelse(gene %in% card,
#                        yes = 'card',
#                        no = NA)) %>%
#   select(-gene, -genome) %>%
#   filter(!is.na(cazy) | !is.na(card)) %>%
#   group_by(OrthoGroup) %>%
#   mutate(annotation = paste(na.omit(c(unique(cazy), unique(card))), collapse = "_")) %>%
#   ungroup()  %>%
#   select(-cazy, -card) %>%
#   distinct() %>%
#   data.frame(., stringsAsFactors = FALSE)
#   
# 
# 
# 
# 
# 
# rownames(ortho_full) <- ortho_full$OrthoGroup
# ortho_full$OrthoGroup <- NULL
# 
# 
# ortho <- read_csv('annotation/fastortho/orthologues_presence_absence.csv')
# ortho_mat <- data.frame(ortho, stringsAsFactors = FALSE)
# rownames(ortho_mat) <- ortho$OrthoGroup
# ortho_mat$OrthoGroup <- NULL
# ortho_mat <- as.matrix(ortho_mat)
# 
# 
# ortho_mat[ortho_mat > 1] <- 1
# 
# 
# breaks = seq(0, 1)
# 
# #colors_f <- colorRampPalette(colors = c('white', 'black'))
# colors <- c('white', 'black')
# 
# 
# annot_colors <- list(annotation = c('card' =  "#1B9E77", 'cazy' = "#D95F02", "cazy_card" = "#7570B3"))
# 
# pheatmap(ortho_mat, show_rownames = FALSE, color = colors, cellwidth = 10, annotation_row = ortho_full, annotation_colors = annot_colors, 
#          annotation_names_row = FALSE, filename = 'plots/orthologous_gene_matrix.pdf')
# 
# 
```



```{r lostEnrichment_not_ffa}



library(topGO)

metadata <- read_csv('tables/metadata.csv') %>% 
  filter(!is.na(genome_id))


## load('rdata/orthologues_long.rdata')

# all_ortho <- orthog_long %>% 
#   dplyr::rename('gene' = genes) %>%
#   filter(genome != "GCA_000225605.1_CmilitarisCM01_v01_protein", !is.na(gene)) %>%
#   mutate(genome = sub('\\..*$', '', genome))
# rm(orthog_long)

uniq_genomes = unique(ortho_full$genome_id) %>% na.omit()
id_map         <- sapply(unique(annot$acc), grep, x = uniq_genomes, value = TRUE)

id_map1        <- names(id_map)
names(id_map1) <- id_map


## make sure all of the genomes are in the annotation (will fail if not)
lapply(uniq_genomes, function(x){
  print(x)
  id_map1[[x]]
})



all_ortho <- ortho_full %>%
  rowwise() %>%
  mutate(genome = id_map1[[genome_id]]) %>%
  group_by(genome) %>%
  mutate(acc = grep(genome, metadata$acc, value = TRUE)) %>%
  ungroup()#%>%
  # left_join(., metadata, by = 'acc')


joined_ortho <- all_ortho %>%
  left_join(annot, by = c('acc', 'gene'))


#dplyr::filter(joined_ortho, tool %in% c('LED', 'MEROPS', 'cazy'), ffa_count == 0) %>% 
#  .$ortho %>% unique() 
  

annot_sum <- joined_ortho %>%
  filter(tool == 'GO') 

annot_sum$Orthogroup %>% n_distinct()

all_out_t <- unique(c(listInput$`Hypomyces_Cladobotryum`, listInput$Trichoderma))

no_out <- listInput$Escovopsis[which(!listInput$Escovopsis %in% all_out_t)]


length(no_out) ## only escovopsis 404


not_ffa <- all_out_t[which(!all_out_t %in% listInput$Escovopsis)]
length(not_ffa) ## lost in escovopsis 1507








ortho2go_df <- annot_sum %>% 
  dplyr::select(Orthogroup, annot) %>%
  group_by(Orthogroup) %>%
  mutate(all_go = paste(unique(annot), collapse = ",")) %>%
  dplyr::select(-annot) %>%
  distinct() #%>%
  #mutate(all_go = paste(unique(strsplit(all_go, ",")[[1]]), collapse = ","))


ortho2go        <- ortho2go_df$all_go
names(ortho2go) <- ortho2go_df$Orthogroup

ortho2go <- lapply(ortho2go, function(x){strsplit(x, ',')[[1]]})

geneNames <- names(ortho2go)
head(geneNames)



## define the genes that I'm interested in (those not in fungus farming ant escovopsis)
myInterestingGenes <- not_ffa
geneList           <- factor(as.integer(geneNames %in% myInterestingGenes))
names(geneList)    <- geneNames

xx <- as.list(GOTERM)

## perform an enrichment test on the genes that are not in FFA
enrich <- lapply(c('BP', 'MF', 'CC'), function(x){
  GOdata <- new("topGOdata", ontology = x, allGenes = geneList,
                annot = annFUN.gene2GO, gene2GO = ortho2go)
  test.stat <- new("classicCount", 
                   testStatistic = GOFisherTest, 
                   name = "Fisher test")
  resultFisher         <- getSigGroups(GOdata, test.stat)
  pVal                 <- data.frame(pval=signif(score(resultFisher), 6),BH.correction=seq(1, length(score(resultFisher))))
  pVal                 <- pVal[order(pVal$pval),]
  pVal$BH.correction   <- signif(p.adjust(pVal$pval, method="BH"), 6)
  pVal.sub             <- pVal[pVal$BH.correction<0.01,]
  pVal.sub$go_id       <- rownames(pVal.sub)
  pVal.try <- cbind(pVal.sub, Term=sapply(pVal.sub$go_id, FUN=function(n){Term(xx[[n]])}), Ontol=sapply(pVal.sub$go_id, FUN=function(n){Ontology(xx[[n]])}))
  
  if (nrow(pVal.try) >0){
    df2                  <- pVal.try[,c("go_id", "Term", "Ontol", "pval", "BH.correction")]
    out <- list(df2, GO, resultFisher)
    invisible(out)
  } else {
    message("No enrichment")
  }
})



all_enrich_df <- bind_rows(enrich[[1]][1], enrich[[2]][1], enrich[[3]][1])
all_enrich_df %>% dplyr::select(go_id, BH.correction, everything()) %>% write_tsv('tables/ffa_reduction_GO.tsv')


## clades shared with escovopsis
5954+486+276+133


## shared across escovopsis
31+2+7+4+25+2+2+18+4+2+4+14+15+67+79

```



```{r revigo_not_ffa}
#printGraph(GOdata, resultFisher, firstSigNodes = nrow(df2), fn.prefix = "tGO", useInfo = "all", pdfSW = TRUE)
#View(df2)


# A plotting R script produced by the REVIGO server at http://revigo.irb.hr/
# If you found REVIGO useful in your work, please cite the following reference:
# Supek F et al. "REVIGO summarizes and visualizes long lists of Gene Ontology
# terms" PLoS ONE 2011. doi:10.1371/journal.pone.0021800


# --------------------------------------------------------------------------
# If you don't have the ggplot2 package installed, uncomment the following line:
# install.packages( "ggplot2" );
library( ggplot2 );
# --------------------------------------------------------------------------
# If you don't have the scales package installed, uncomment the following line:
# install.packages( "scales" );
library( scales );


# --------------------------------------------------------------------------
# A plotting R script produced by the REVIGO server at http://revigo.irb.hr/
# If you found REVIGO useful in your work, please cite the following reference:
# Supek F et al. "REVIGO summarizes and visualizes long lists of Gene Ontology
# terms" PLoS ONE 2011. doi:10.1371/journal.pone.0021800

# --------------------------------------------------------------------------
# If you don't have the ggplot2 package installed, uncomment the following line:
# install.packages( "ggplot2" );
library( ggplot2 );

# --------------------------------------------------------------------------
# If you don't have the scales package installed, uncomment the following line:
# install.packages( "scales" );
library( scales );

# --------------------------------------------------------------------------
# Here is your data from REVIGO. Scroll down for plot configuration options.

revigo.names <- c("term_ID","description","frequency_%","plot_X","plot_Y","plot_size","value","uniqueness","dispensability");
revigo.data <- rbind(c("GO:0006351","transcription, DNA-templated",2.021,-4.888,4.215,5.658,-4.205,0.838,0.000),
c("GO:0006355","regulation of transcription, DNA-templated",9.955,4.055,5.165,6.350,-4.205,0.159,0.000),
c("GO:0055114","oxidation-reduction process",100.000,-3.005,-4.622,6.555,-7.465,1.000,0.000),
c("GO:0055085","transmembrane transport",13.277,3.967,-3.936,6.475,-5.388,0.996,0.011),
c("GO:0019222","regulation of metabolic process",13.187,4.624,4.212,6.472,-2.447,0.326,0.566),
c("GO:0032774","RNA biosynthetic process",2.256,-5.212,3.304,5.705,-4.158,0.867,0.649));

one.data <- data.frame(revigo.data);
names(one.data) <- revigo.names;
one.data <- one.data [(one.data$plot_X != "null" & one.data$plot_Y != "null"), ];
one.data$plot_X <- as.numeric( as.character(one.data$plot_X) );
one.data$plot_Y <- as.numeric( as.character(one.data$plot_Y) );
one.data$plot_size <- as.numeric( as.character(one.data$plot_size) );
one.data$value <- as.numeric( as.character(one.data$value) );
one.data$frequency <- as.numeric( as.character(one.data$frequency) );
one.data$uniqueness <- as.numeric( as.character(one.data$uniqueness) );
one.data$dispensability <- as.numeric( as.character(one.data$dispensability) );
#head(one.data);


# --------------------------------------------------------------------------
# Names of the axes, sizes of the numbers and letters, names of the columns,
# etc. can be changed below

p1 <- ggplot( data = one.data );
p1 <- p1 + geom_point( aes( plot_X, plot_Y, colour = value, size = plot_size), alpha = I(0.6) ) + scale_size_area();
p1 <- p1 + scale_colour_gradientn( colours = c("blue", "green", "yellow", "red"), limits = c( min(one.data$value), 0) );

p1 <- p1 + geom_point( aes(plot_X, plot_Y, size = plot_size), shape = 21, fill = "transparent", colour = I (alpha ("black", 0.6) )) + scale_size_area();

p1 <- p1 + scale_size( range=c(5, 30)) + theme_bw(); # + scale_fill_gradientn(colours = heat_hcl(7), limits = c(-300, 0) );

ex <- one.data #[ one.data$dispensability < 0.15, ];

p1 <- p1 + geom_text( data = ex, aes(plot_X, plot_Y, label = description), colour = I(alpha("black", 0.85)), size = 3 );
p1 <- p1 + labs (y = "semantic space x", x = "semantic space y");
p1 <- p1 + theme(legend.key = element_blank()) ;
one.x_range = max(one.data$plot_X) - min(one.data$plot_X);
one.y_range = max(one.data$plot_Y) - min(one.data$plot_Y);
p1 <- p1 + xlim(min(one.data$plot_X)-one.x_range/10,max(one.data$plot_X)+one.x_range/10);
p1 <- p1 + ylim(min(one.data$plot_Y)-one.y_range/10,max(one.data$plot_Y)+one.y_range/10);


# --------------------------------------------------------------------------
# Output the plot to screen

p1;

# Uncomment the line below to also save the plot to a file.
# The file type depends on the extension (default=pdf).

# ggsave("/path_to_your_file/revigo-plot.pdf");



# --------------------------------------------------------------------------
# Output the plot to screen

p1 + ggsave('plots/reduced_all_revigo.pdf', width = 7, height = 7)






```

